libdevcrypto/LibSnark.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ This file is part of cpp-ethereum.
+
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file LibSnark.cpp
+ */
+
+#include <libdevcrypto/LibSnark.h>
+
+#include <algebra/curves/alt_bn128/alt_bn128_g1.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_g2.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pairing.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pp.hpp>
+
+#include <libdevcrypto/Exceptions.h>
+
+#include <libdevcore/CommonIO.h>
+#include <libdevcore/Log.h>
+
+#include <mutex>
+
+using namespace std;
+using namespace dev;
+using namespace dev::crypto;
+
+namespace
+{
+
+std::once_flag initLibSnarkFlag;
+
+void initLibSnark()
+{
+	call_once(initLibSnarkFlag,
+		[](){
+			cnote << "calling libff::alt_bn128_pp::init_public_params().";
+			libff::alt_bn128_pp::init_public_params();
+		});
+}
+
+libff::bigint<libff::alt_bn128_q_limbs> toLibsnarkBigint(h256 const& _x)
+{
+	libff::bigint<libff::alt_bn128_q_limbs> x;
+	auto const N = x.N;
+	constexpr auto L = sizeof(x.data[0]);
+	for (unsigned i = 0; i < N; i++)
+		for (unsigned j = 0; j < L; j++)
+			x.data[N - 1 - i] |= uint64_t(_x[i * L + j]) << (8 * (L - 1 - j));
```

Comment:
This might be a problem when `x.data[0]` is wider than 64bit, but I'm not sure if we should go for
```
            x.data[N - 1 - i] |= (std::remove_reference<decltype(x.data[0])>::type)(_x[i * L + j]) << (8 * (L - 1 - j));
```